File: AILang/benchmarks/c/gauntlet_ownership_erased.c
#include <stdint.h>
#include <stdio.h>

static inline int64_t decimal_len_i64(int64_t value) {
    uint64_t v = value < 0 ? (uint64_t)(-value) : (uint64_t)value;
    int64_t len = value < 0 ? 1 : 0;
    do {
        ++len;
        v /= 10u;
    } while (v != 0u);
    return len;
}

static int64_t ownership_churn(int64_t iterations) {
    int64_t acc = 0;
    for (int64_t i = 0; i < iterations; ++i) {
        int64_t seed = i % 97LL;
        int64_t score = 4LL + decimal_len_i64(i) + seed + (seed + 1LL) + (seed + 2LL);
        acc = (acc + score) % 1000000007LL;
    }
    return acc;
}

int main(void) {
    printf("%lld\n", (long long)ownership_churn(8000000LL));
    return 0;
}

File: AILang/benchmarks/rust/gauntlet_ownership_erased.rs
#[inline(always)]
fn decimal_len_i64(value: i64) -> i64 {
    let mut v = if value < 0 {
        value.wrapping_neg() as u64
    } else {
        value as u64
    };
    let mut len = if value < 0 { 1_i64 } else { 0_i64 };
    loop {
        len += 1;
        v /= 10;
        if v == 0 {
            return len;
        }
    }
}

fn ownership_churn(iterations: i64) -> i64 {
    let mut acc = 0_i64;
    let mut i = 0_i64;
    while i < iterations {
        let seed = i % 97;
        let score = 4 + decimal_len_i64(i) + seed + (seed + 1) + (seed + 2);
        acc = (acc + score) % 1_000_000_007;
        i += 1;
    }
    acc
}

fn main() {
    println!("{}", ownership_churn(8_000_000));
}
